File: python/modcdp/transport/UpstreamTransport.py
# MODCDP_TRANSLATE: KEEP THIS FILE TRANSLATED ACROSS TYPESCRIPT, PYTHON, AND GO.
# Keep all shapes, signatures, behavior, and tests 1:1 in sync with:
# - ./js/src/transport/UpstreamTransport.ts
# - ./go/modcdp/transport/UpstreamTransport.go
from __future__ import annotations

import json
import threading
from collections.abc import Callable, Mapping
from queue import Empty, Queue
from typing import Any, Literal, TypeAlias, overload

from pydantic import BaseModel

from ..types.modcdp import ModCDPUpstreamConfig, ProtocolPayload, ProtocolResult, _isObjectMap
from ..types.toJSON import modCDPToJSON


UpstreamMode: TypeAlias = Literal["ws"]
UpstreamPeerKind: TypeAlias = Literal["browser_cdp", "modcdp_server"]
UpstreamTransportConfig: TypeAlias = ModCDPUpstreamConfig


class UpstreamTransport:
    upstream_mode: UpstreamMode = "ws"
    peer_kind: UpstreamPeerKind = "browser_cdp"
    url: str | None = None

    def __init__(self, config: UpstreamTransportConfig | dict[str, Any] | None = None) -> None:
        self.config = _upstream_transport_config(config)
        self._next_id = 0
        self._pending: dict[int, tuple[str, Queue[dict[str, Any]]]] = {}
        self._lock = threading.Lock()
        self._recv_listeners: list[Callable[[dict[str, Any]], None]] = []
        self._close_listeners: list[Callable[[Exception], None]] = []
        self._event_listeners: dict[str, list[Callable[[dict[str, Any], str | None, str | None], None]]] = {}

    def connect(self) -> None:
        raise NotImplementedError(f"{type(self).__name__}.connect is not implemented.")

    def update(self, config: UpstreamTransportConfig | dict[str, Any] | None = None) -> "UpstreamTransport":
        incoming = _upstream_transport_config(config)
        self.config = UpstreamTransportConfig.model_validate({**self.config.model_dump(), **incoming.model_dump(exclude_unset=True)})
        return self

    def configForLauncher(self) -> dict[str, Any]:
        return {}

    def close(self) -> None:
        return None

    @overload
    def send(
        self,
        command: str,
        params: ProtocolPayload | None = None,
        session_id: str | None = None,
        *,
        timeout_ms: int | None = None,
    ) -> ProtocolResult: ...

    @overload
    def send(
        self,
        command: dict[str, Any],
        params: ProtocolPayload | None = None,
        session_id: str | None = None,
        *,
        timeout_ms: int | None = None,
    ) -> None: ...

    def send(
        self,
        command: dict[str, Any] | str | object,
        params: ProtocolPayload | None = None,
        session_id: str | None = None,
        *,
        timeout_ms: int | None = None,
    ) -> ProtocolResult | None:
        if isinstance(command, dict):
            raise NotImplementedError(f"{type(self).__name__}.send is not implemented.")
        method = _cdp_name(command)
        effective_timeout_ms = timeout_ms if timeout_ms is not None else self.config.upstream_cdp_send_timeout_ms
        with self._lock:
            self._next_id += 1
            msg_id = self._next_id
            done: Queue[dict[str, Any]] = Queue()
            self._pending[msg_id] = (method, done)
        message: dict[str, Any] = {"id": msg_id, "method": method, "params": params or {}}
        if session_id:
            message["sessionId"] = session_id
        try:
            self.send(message)
        except Exception:
            with self._lock:
                self._pending.pop(msg_id, None)
            raise
        try:
            response = done.get(timeout=effective_timeout_ms / 1000 if effective_timeout_ms > 0 else None)
        except Empty:
            with self._lock:
                self._pending.pop(msg_id, None)
            raise RuntimeError(f"{method} timed out after {effective_timeout_ms}ms")
        err = response.get("error")
        if err:
            raise RuntimeError(f"{method} failed: {err.get('message', err) if isinstance(err, dict) else err}")
        result = response.get("result")
        return result if isinstance(result, dict) else {}

    def onRecv(self, listener: Callable[[dict[str, Any]], None]) -> Callable[[], None]:
        self._recv_listeners.append(listener)

        removed = False

        def stop() -> None:
            nonlocal removed
            if removed:
                return
            removed = True
            try:
                self._recv_listeners.remove(listener)
            except ValueError:
                return

        return stop

    def onClose(self, listener: Callable[[Exception], None]) -> Callable[[], None]:
        self._close_listeners.append(listener)

        removed = False

        def stop() -> None:
            nonlocal removed
            if removed:
                return
            removed = True
            try:
                self._close_listeners.remove(listener)
            except ValueError:
                return

        return stop

    def on(self, event: Any, listener: Callable[[dict[str, Any], str | None, str | None], None]) -> Callable[[], None]:
        event_name = _cdp_event_name(event)

        def typed_listener(payload: dict[str, Any], target_id: str | None, session_id: str | None) -> None:
            listener(_parse_event_payload(event, payload), target_id, session_id)

        listeners = self._event_listeners.setdefault(event_name, [])
        listeners.append(typed_listener)

        removed = False

        def stop() -> None:
            nonlocal removed
            if removed:
                return
            removed = True
            current_listeners = self._event_listeners.get(event_name)
            if current_listeners is None:
                return
            try:
                current_listeners.remove(typed_listener)
            except ValueError:
                return
            if not current_listeners:
                self._event_listeners.pop(event_name, None)

        return stop

    def getTargets(self) -> list[dict[str, Any]]:
        result = self.send("Target.getTargets", {})
        target_infos = result.get("targetInfos") if isinstance(result, dict) else None
        return [target for target in target_infos if _isObjectMap(target)] if isinstance(target_infos, list) else []

    def resolveTargetId(self, params: dict[str, Any] | None = None) -> str | None:
        target_id = (params or {}).get("targetId")
        return target_id if isinstance(target_id, str) and target_id else None

    def createTarget(self, url: str) -> str:
        result = self.send("Target.createTarget", {"url": url})
        target_id = result.get("targetId") if isinstance(result, dict) else None
        if not isinstance(target_id, str) or not target_id:
            raise RuntimeError("Target.createTarget returned no targetId")
        return target_id

    def attachToTarget(self, target_id: str) -> str | None:
        result = self.send("Target.attachToTarget", {"targetId": target_id, "flatten": True})
        session_id = result.get("sessionId") if isinstance(result, dict) else None
        return session_id if isinstance(session_id, str) and session_id else None

    def detachFromTarget(self, session_id: str) -> None:
        self.send("Target.detachFromTarget", {"sessionId": session_id})

    def waitForPeer(self, config: dict[str, Any] | None = None) -> None:
        return None

    def toJSON(self) -> dict[str, object]:
        config = self.config.model_dump(mode="json")
        return modCDPToJSON(
            self,
            {
                "config": config,
                "state": {
                    "pending": len(self._pending),
                    "recv_listeners": len(self._recv_listeners),
                    "close_listeners": len(self._close_listeners),
                    "event_listeners": len(self._event_listeners),
                }
            },
        )

    def _emit_recv(self, message: dict[str, Any]) -> None:
        for listener in list(self._recv_listeners):
            listener(message)

    def _emit_close(self, error: Exception) -> None:
        self._settle_pending(error)
        for listener in list(self._close_listeners):
            listener(error)

    def _settle_pending(self, error: Exception) -> None:
        with self._lock:
            pending = list(self._pending.values())
            self._pending.clear()
        for _, done in pending:
            done.put({"error": {"message": str(error)}})

    def _parse_and_emit_recv(self, data: str | bytes) -> None:
        raw = data.decode() if isinstance(data, bytes) else data
        parsed = json.loads(raw)
        if not isinstance(parsed, dict):
            return
        if isinstance(parsed.get("id"), int):
            with self._lock:
                entry = self._pending.pop(parsed["id"], None)
            if entry:
                entry[1].put(parsed)
            self._emit_recv(parsed)
            return
        method = parsed.get("method")
        params = parsed.get("params")
        session_id = parsed.get("sessionId")
        if isinstance(method, str):
            event_params = params if isinstance(params, dict) else {}
            self._emit_upstream_event(method, event_params, None, session_id if isinstance(session_id, str) else None)
        self._emit_recv(parsed)

    def _emit_upstream_event(
        self,
        method: str,
        payload: dict[str, Any],
        target_id: str | None,
        session_id: str | None,
    ) -> None:
        for listener in list(self._event_listeners.get(method, [])):
            listener(payload, target_id, session_id)
def _upstream_transport_config(config: UpstreamTransportConfig | dict[str, Any] | None = None) -> UpstreamTransportConfig:
    if isinstance(config, UpstreamTransportConfig):
        return config
    return UpstreamTransportConfig.model_validate(config or {})


def _cdp_name(command: object) -> str:
    if isinstance(command, str):
        return command
    meta_fn = getattr(command, "meta", None)
    meta = meta_fn() if callable(meta_fn) else None
    candidates = (
        getattr(command, "cdp_command_name", None),
        getattr(command, "id", None),
        getattr(meta, "cdp_command_name", None) if meta is not None else None,
        meta.get("cdp_command_name") if isinstance(meta, Mapping) else None,
        meta.get("id") if isinstance(meta, Mapping) else None,
        getattr(command, "name", None),
    )
    name = next((candidate for candidate in candidates if isinstance(candidate, str) and candidate), None)
    if name is None:
        raise TypeError("command must be a CDP method string or generated command object")
    return name


def _cdp_event_name(event: object) -> str:
    if isinstance(event, str):
        return event
    meta_fn = getattr(event, "meta", None)
    meta = meta_fn() if callable(meta_fn) else None
    candidates = (
        getattr(event, "cdp_event_name", None),
        getattr(meta, "cdp_event_name", None) if meta is not None else None,
        meta.get("cdp_event_name") if isinstance(meta, Mapping) else None,
        getattr(event, "id", None),
        getattr(meta, "id", None) if meta is not None else None,
        meta.get("id") if isinstance(meta, Mapping) else None,
        getattr(event, "name", None),
    )
    name = next((candidate for candidate in candidates if isinstance(candidate, str) and candidate), None)
    if name is None:
        raise TypeError("event must be a CDP event name string or generated event object")
    return name


def _parse_event_payload(event: object, payload: dict[str, Any]) -> dict[str, Any]:
    if isinstance(event, type) and issubclass(event, BaseModel):
        parsed = event.model_validate(payload)
        return parsed.model_dump(mode="json", exclude_none=True, by_alias=True)
    return payload

File: go/modcdp/transport/UpstreamTransport.go
// MODCDP_TRANSLATE: KEEP THIS FILE TRANSLATED ACROSS TYPESCRIPT, PYTHON, AND GO.
// Keep all shapes, signatures, behavior, and tests 1:1 in sync with:
// - ./js/src/transport/UpstreamTransport.ts
// - ./python/modcdp/transport/UpstreamTransport.py
package transport

import (
	"encoding/json"
	"fmt"
	"net"
	"net/url"
	"strconv"
	"strings"
	"sync"
	"time"

	"github.com/browserbase/modcdp/go/modcdp/injector"
	"github.com/browserbase/modcdp/go/modcdp/types"
)

type InjectorConfig = types.InjectorConfig
type LauncherConfig = types.LauncherConfig
type UpstreamTransportConfig = types.UpstreamTransportConfig

const DefaultModCDPExtensionID = injector.DefaultModCDPExtensionID

func firstNonEmptyString(values ...string) string {
	for _, value := range values {
		if value != "" {
			return value
		}
	}
	return ""
}

func freePort() (int, error) {
	listener, err := net.Listen("tcp", "127.0.0.1:0")
	if err != nil {
		return 0, err
	}
	defer listener.Close()
	return listener.Addr().(*net.TCPAddr).Port, nil
}

type UpstreamMode string
type UpstreamPeerKind string

const (
	UpstreamModeWS UpstreamMode = "ws"

	UpstreamPeerKindBrowserCDP   UpstreamPeerKind = "browser_cdp"
	UpstreamPeerKindModCDPServer UpstreamPeerKind = "modcdp_server"
)

type HostPort struct {
	Host string `json:"host"`
	Port int    `json:"port"`
}

type UpstreamTransport struct {
	Config         UpstreamTransportConfig
	PeerKind       UpstreamPeerKind
	recvListeners  []recvListener
	closeListeners []closeListener
	eventListeners map[string][]upstreamEventListener
	listenerMu     sync.Mutex
	nextListenerID int64
	nextID         int64
	pending        map[int64]chan map[string]any
	pendingMu      sync.Mutex
	writeCommand   func(map[string]any) error
}

type recvListener struct {
	id int64
	fn func(map[string]any)
}

type closeListener struct {
	id int64
	fn func(error)
}

type upstreamEventListener struct {
	id int64
	fn func(map[string]any, string, string)
}

func NewUpstreamTransport(config UpstreamTransportConfig) UpstreamTransport {
	if config.UpstreamMode == "" {
		config.UpstreamMode = string(UpstreamModeWS)
	}
	if config.UpstreamWSConnectErrorSettleTimeoutMS == 0 {
		config.UpstreamWSConnectErrorSettleTimeoutMS = 250
	}
	if config.UpstreamCDPSendTimeoutMS == 0 {
		config.UpstreamCDPSendTimeoutMS = 10_000
	}
	return UpstreamTransport{
		Config:         config,
		PeerKind:       UpstreamPeerKindBrowserCDP,
		eventListeners: map[string][]upstreamEventListener{},
		pending:        map[int64]chan map[string]any{},
		writeCommand: func(map[string]any) error {
			return fmt.Errorf("UpstreamTransport.send is not implemented")
		},
	}
}

func (e *UpstreamTransport) Update(config map[string]any) {
	if config == nil {
		return
	}
	if value, ok := config["upstream_ws_cdp_url"].(string); ok {
		e.Config.UpstreamWSCDPURL = value
	}
	if value, ok := config["upstream_mode"].(string); ok && value != "" {
		e.Config.UpstreamMode = value
	}
	if value, ok := intFromConfig(config["upstream_ws_connect_error_settle_timeout_ms"]); ok {
		e.Config.UpstreamWSConnectErrorSettleTimeoutMS = value
	}
	if value, ok := intFromConfig(config["upstream_cdp_send_timeout_ms"]); ok {
		e.Config.UpstreamCDPSendTimeoutMS = value
	}
}

func (e *UpstreamTransport) Connect() error {
	return fmt.Errorf("%T.Connect is not implemented", e)
}

func (e *UpstreamTransport) Close() error {
	return nil
}

func (e *UpstreamTransport) Send(command any, params map[string]any, sessionID string, timeout ...time.Duration) (map[string]any, error) {
	method, ok := command.(string)
	if !ok {
		message, ok := command.(types.CdpCommandMessage)
		if !ok {
			return nil, fmt.Errorf("command must be a CDP method name or CdpCommandMessage")
		}
		payload := map[string]any{
			"id":     message.ID,
			"method": message.Method,
		}
		if message.Params != nil {
			payload["params"] = message.Params
		}
		if message.SessionID != "" {
			payload["sessionId"] = message.SessionID
		}
		return map[string]any{}, e.writeCommand(payload)
	}
	e.pendingMu.Lock()
	e.nextID++
	id := e.nextID
	done := make(chan map[string]any, 1)
	e.pending[id] = done
	e.pendingMu.Unlock()

	message := map[string]any{"id": id, "method": method, "params": params}
	if sessionID != "" {
		message["sessionId"] = sessionID
	}
	if err := e.writeCommand(message); err != nil {
		e.pendingMu.Lock()
		delete(e.pending, id)
		e.pendingMu.Unlock()
		return nil, err
	}
	effectiveTimeout := time.Duration(e.Config.UpstreamCDPSendTimeoutMS) * time.Millisecond
	if len(timeout) > 0 {
		effectiveTimeout = timeout[0]
	}
	if effectiveTimeout <= 0 {
		response := <-done
		if errObj, ok := response["error"].(map[string]any); ok {
			return nil, fmt.Errorf("%s failed: %v", method, errObj["message"])
		}
		if result, ok := response["result"].(map[string]any); ok {
			return result, nil
		}
		return map[string]any{}, nil
	}
	select {
	case <-time.After(effectiveTimeout):
		e.pendingMu.Lock()
		delete(e.pending, id)
		e.pendingMu.Unlock()
		return nil, fmt.Errorf("%s timed out after %s", method, effectiveTimeout)
	case response := <-done:
		if errObj, ok := response["error"].(map[string]any); ok {
			return nil, fmt.Errorf("%s failed: %v", method, errObj["message"])
		}
		if result, ok := response["result"].(map[string]any); ok {
			return result, nil
		}
		return map[string]any{}, nil
	}
}

func (e *UpstreamTransport) ConfigForLauncher() LauncherConfig {
	return LauncherConfig{}
}

func (e *UpstreamTransport) GetTargets() ([]map[string]any, error) {
	result, err := e.Send("Target.getTargets", map[string]any{}, "")
	if err != nil {
		return nil, err
	}
	targetInfos, _ := result["targetInfos"].([]any)
	targets := []map[string]any{}
	for _, targetInfo := range targetInfos {
		target, _ := targetInfo.(map[string]any)
		targets = append(targets, target)
	}
	return targets, nil
}

func (e *UpstreamTransport) ResolveTargetID(params map[string]any) string {
	targetID, _ := params["targetId"].(string)
	return targetID
}

func (e *UpstreamTransport) CreateTarget(url string) (string, error) {
	result, err := e.Send("Target.createTarget", map[string]any{"url": url}, "")
	if err != nil {
		return "", err
	}
	targetID, _ := result["targetId"].(string)
	if targetID == "" {
		return "", fmt.Errorf("Target.createTarget returned no targetId")
	}
	return targetID, nil
}

func (e *UpstreamTransport) AttachToTarget(targetID string) (string, error) {
	result, err := e.Send("Target.attachToTarget", map[string]any{"targetId": targetID, "flatten": true}, "")
	if err != nil {
		return "", err
	}
	sessionID, _ := result["sessionId"].(string)
	return sessionID, nil
}

func (e *UpstreamTransport) DetachFromTarget(sessionID string) error {
	_, err := e.Send("Target.detachFromTarget", map[string]any{"sessionId": sessionID}, "")
	return err
}

func (e *UpstreamTransport) OnRecv(listener func(map[string]any)) func() {
	e.listenerMu.Lock()
	e.nextListenerID++
	id := e.nextListenerID
	e.recvListeners = append(e.recvListeners, recvListener{id: id, fn: listener})
	e.listenerMu.Unlock()
	var once sync.Once
	return func() {
		once.Do(func() {
			e.listenerMu.Lock()
			defer e.listenerMu.Unlock()
			for index, candidate := range e.recvListeners {
				if candidate.id != id {
					continue
				}
				e.recvListeners = append(e.recvListeners[:index], e.recvListeners[index+1:]...)
				return
			}
		})
	}
}

func (e *UpstreamTransport) OnClose(listener func(error)) func() {
	e.listenerMu.Lock()
	e.nextListenerID++
	id := e.nextListenerID
	e.closeListeners = append(e.closeListeners, closeListener{id: id, fn: listener})
	e.listenerMu.Unlock()
	var once sync.Once
	return func() {
		once.Do(func() {
			e.listenerMu.Lock()
			defer e.listenerMu.Unlock()
			for index, candidate := range e.closeListeners {
				if candidate.id != id {
					continue
				}
				e.closeListeners = append(e.closeListeners[:index], e.closeListeners[index+1:]...)
				return
			}
		})
	}
}

func (e *UpstreamTransport) On(event string, listener func(map[string]any, string, string)) func() {
	e.listenerMu.Lock()
	if e.eventListeners == nil {
		e.eventListeners = map[string][]upstreamEventListener{}
	}
	e.nextListenerID++
	id := e.nextListenerID
	e.eventListeners[event] = append(e.eventListeners[event], upstreamEventListener{id: id, fn: listener})
	e.listenerMu.Unlock()
	var once sync.Once
	return func() {
		once.Do(func() {
			e.listenerMu.Lock()
			defer e.listenerMu.Unlock()
			listeners := e.eventListeners[event]
			for index, candidate := range listeners {
				if candidate.id != id {
					continue
				}
				listeners = append(listeners[:index], listeners[index+1:]...)
				if len(listeners) == 0 {
					delete(e.eventListeners, event)
				} else {
					e.eventListeners[event] = listeners
				}
				return
			}
		})
	}
}

func (e *UpstreamTransport) emitRecv(message map[string]any) {
	if id, ok := commandID(message["id"]); ok {
		e.pendingMu.Lock()
		done := e.pending[id]
		delete(e.pending, id)
		e.pendingMu.Unlock()
		if done != nil {
			done <- message
		}
	}
	if _, ok := commandID(message["id"]); !ok {
		method, _ := message["method"].(string)
		params, _ := message["params"].(map[string]any)
		sessionID, _ := message["sessionId"].(string)
		if method != "" {
			if params == nil {
				params = map[string]any{}
			}
			e.emitUpstreamEvent(method, params, "", sessionID)
		}
	}
	e.listenerMu.Lock()
	listeners := append([]recvListener(nil), e.recvListeners...)
	e.listenerMu.Unlock()
	for _, listener := range listeners {
		listener.fn(message)
	}
}

func (e *UpstreamTransport) EmitRecv(message map[string]any) {
	e.emitRecv(message)
}

func (e *UpstreamTransport) parseAndEmitRecv(data []byte) error {
	var message map[string]any
	if err := json.Unmarshal(data, &message); err != nil {
		return fmt.Errorf("invalid CDP message: %w", err)
	}
	if _, ok := commandID(message["id"]); ok {
		if errObj, hasError := message["error"]; hasError && errObj != nil {
			errorMap, ok := errObj.(map[string]any)
			if !ok {
				return fmt.Errorf("invalid CDP response error")
			}
			if message, ok := errorMap["message"].(string); !ok || message == "" {
				return fmt.Errorf("invalid CDP response error message")
			}
		}
		if sessionID, ok := message["sessionId"]; ok && sessionID != nil {
			if _, ok := sessionID.(string); !ok {
				return fmt.Errorf("invalid CDP response sessionId")
			}
		}
		e.emitRecv(message)
		return nil
	}
	if _, hasID := message["id"]; hasID {
		return fmt.Errorf("invalid CDP response id")
	}
	method, _ := message["method"].(string)
	if method == "" {
		return fmt.Errorf("invalid CDP event method")
	}
	if params, ok := message["params"]; ok && params != nil {
		if _, ok := params.(map[string]any); !ok {
			return fmt.Errorf("invalid CDP event params")
		}
	}
	if sessionID, ok := message["sessionId"]; ok && sessionID != nil {
		if _, ok := sessionID.(string); !ok {
			return fmt.Errorf("invalid CDP event sessionId")
		}
	}
	e.emitRecv(message)
	return nil
}

func (e *UpstreamTransport) emitUpstreamEvent(method string, payload map[string]any, targetID string, sessionID string) {
	e.listenerMu.Lock()
	listeners := append([]upstreamEventListener(nil), e.eventListeners[method]...)
	e.listenerMu.Unlock()
	for _, listener := range listeners {
		listener.fn(payload, targetID, sessionID)
	}
}

func (e *UpstreamTransport) emitClose(err error) {
	e.pendingMu.Lock()
	pending := e.pending
	e.pending = map[int64]chan map[string]any{}
	e.pendingMu.Unlock()
	for _, done := range pending {
		done <- map[string]any{"error": map[string]any{"message": fmt.Sprintf("connection closed: %v", err)}}
	}
	e.listenerMu.Lock()
	listeners := append([]closeListener(nil), e.closeListeners...)
	e.listenerMu.Unlock()
	for _, listener := range listeners {
		listener.fn(err)
	}
}

func (e *UpstreamTransport) EmitClose(err error) {
	e.emitClose(err)
}

func (e *UpstreamTransport) WaitForPeer() error {
	return nil
}

func (e *UpstreamTransport) PeerGeneration() int64 {
	return 0
}

func (e *UpstreamTransport) ToJSON() map[string]any {
	e.pendingMu.Lock()
	pending := len(e.pending)
	e.pendingMu.Unlock()
	e.listenerMu.Lock()
	recvListeners := len(e.recvListeners)
	closeListeners := len(e.closeListeners)
	eventListeners := len(e.eventListeners)
	e.listenerMu.Unlock()
	config := e.Config
	return types.ModCDPToJSON(e, types.ModCDPJSONConfig{
		Config: config,
		State: map[string]any{
			"pending":         pending,
			"recv_listeners":  recvListeners,
			"close_listeners": closeListeners,
			"event_listeners": eventListeners,
		},
	})
}

func ParseHostPort(value string, defaultHost string, defaultPort int) (HostPort, error) {
	parseValue := value
	if !strings.Contains(value, "://") {
		parseValue = "ws://" + value
	}
	parsed, err := url.Parse(parseValue)
	if err != nil {
		return HostPort{}, err
	}
	host := parsed.Hostname()
	if host == "" {
		host = defaultHost
	}
	port := defaultPort
	if parsed.Port() != "" {
		parsedPort, ok := intFromConfig(parsed.Port())
		if !ok {
			return HostPort{}, fmt.Errorf("Invalid host:port %s", value)
		}
		port = parsedPort
	}
	if port <= 0 || port > 65_535 {
		return HostPort{}, fmt.Errorf("Invalid host:port %s", value)
	}
	return HostPort{Host: host, Port: port}, nil
}

func intFromConfig(value any) (int, bool) {
	switch typed := value.(type) {
	case int:
		return typed, true
	case int64:
		return int(typed), true
	case float64:
		return int(typed), true
	case float32:
		return int(typed), true
	case string:
		parsed, err := strconv.Atoi(typed)
		if err != nil {
			return 0, false
		}
		return parsed, true
	default:
		return 0, false
	}
}

func commandID(value any) (int64, bool) {
	switch typed := value.(type) {
	case int:
		return int64(typed), true
	case int64:
		return typed, true
	case float64:
		return int64(typed), true
	default:
		return 0, false
	}
}
